### PR TITLE
Deprecate support for directly accessing logger

### PR DESCRIPTION
**Pull request by vyasr**
## Description

Contributes to https://github.com/rapidsai/build-planning/issues/104

This PR removes support for accessing rmm's underlying spdlog logger directly.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: include/rmm/logger.hpp [hunk 1]
```diff
@@ -107,7 +107,10 @@ struct bytes {
  *
  * @return spdlog::logger& The logger.
  */
-RMM_EXPORT inline spdlog::logger& logger()
+[[deprecated(
+  "Support for direct access to spdlog loggers in rmm is planned for "
+  "removal")]] RMM_EXPORT inline spdlog::logger&
+logger()
```
Comment (wence-) [inline include/rmm/logger.hpp:113]: Note this makes all the `RMM_LOG_XXX` macros produce warnings, I think.

Comment (vyasr) [reply include/rmm/logger.hpp:113]: Good point. I can add a level of indirection to help with this.

**Review by bdice — APPROVED** [review]

**Review by wence- — APPROVED** [review]

**Comment by wence-** [comment]
Setting to do not merge, I think the Cython bindings in `rmm/_lib/logger.pyx` also want updated.

File: include/rmm/logger.hpp [hunk 2]
```diff
@@ -96,6 +96,11 @@ struct bytes {
   }
 };
 
+inline spdlog::logger& logger()
```
Comment (harrism) [inline include/rmm/logger.hpp:99]: It's still possible for users to call this function even though it's in `detail`. So how does this help with simplifying dependencies?

Comment (wence-) [reply include/rmm/logger.hpp:99]: We no longer leak an ABI-unstable third party object out of the public api

**Review by harrism — APPROVED** [review]

**Comment by wence-** [comment]
@vyasr am I right in my belief that the cython bindings now also want to use the detail API call?

File: include/rmm/logger.hpp (hunk 2 above)
Comment (vyasr) [reply include/rmm/logger.hpp:99]: Yes, exactly. This function was primarily added so as to have a way to call this internally in places that need it for the moment without getting compiler warnings.

**Comment by vyasr** [comment]
> @vyasr am I right in my belief that the cython bindings now also want to use the detail API call?

You are correct. In fact we probably need to do more than that. This PR can probably just do that, but we will need to add some public rmm APIs to do things that were previously being done by accessing the spdlog logger's APIs directly in the Cython layer. I'll work on that in a follow-up where I'll try to isolate the core APIs that rmm will need to support in order to not expose spdlog directly. I suspect that we're going to have to ask users to do some sort of unique symbol generation on their end so as to be able to generate a private set of spdlog symbols that they don't leak to consumers.

**Review by wence- — APPROVED** [review]
Thanks!

**Comment by harrism** [comment]
Reminder we need an issue for the required Cython/Python changes.

**Comment by vyasr** [comment]
https://github.com/rapidsai/rmm/issues/1709

**Pull request merged**